sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -402,11 +402,13 @@ trait Row extends Serializable {
               return false
             }
           case f1: Float if java.lang.Float.isNaN(f1) =>
-            if (!o2.isInstanceOf[Float] || ! java.lang.Float.isNaN(o2.asInstanceOf[Float])) {
+            if (!(o2.isInstanceOf[Float] && java.lang.Float.isNaN(o2.asInstanceOf[Float]) ||
```

Comment:
Thank you for attention, @srowen . I agree with you and @rxin in the current Spark status view point.

Today, when I'm reading [Spark SQL, DataFrames and Datasets Guide: NaN Semantics](http://spark.apache.org/docs/latest/sql-programming-guide.html), I just suddenly want to update this PR.
> **NaN Semantics**
> There is specially handling for not-a-number (NaN) when dealing with float or double types that does not exactly match standard floating point semantics. Specifically:
> - NaN = NaN returns true.

I'm still digging to find some useful cases for this PR outside SQL layers.